subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/SnapshotHierarchy.java

```diff
@@ -47,23 +45,54 @@
      * Returns a hierarchy augmented by the information of the snapshot at the absolute path.
      */
     @CheckReturnValue
-    SnapshotHierarchy store(String absolutePath, MetadataSnapshot snapshot);
+    SnapshotHierarchy store(String absolutePath, MetadataSnapshot snapshot, NodeDiffListener diffListener);
 
     /**
      * Returns a hierarchy without any information at the absolute path.
      */
     @CheckReturnValue
-    SnapshotHierarchy invalidate(String absolutePath);
+    SnapshotHierarchy invalidate(String absolutePath, NodeDiffListener diffListener);
 
     /**
      * The empty hierarchy.
      */
     @CheckReturnValue
     SnapshotHierarchy empty();
 
-    void visitSnapshots(SnapshotVisitor snapshotVisitor);
+    void visitSnapshotRoots(SnapshotVisitor snapshotVisitor);
 
     interface SnapshotVisitor {
-        void visitSnapshot(CompleteFileSystemLocationSnapshot snapshot, boolean rootOfCompleteHierarchy);
+        void visitSnapshotRoot(CompleteFileSystemLocationSnapshot snapshot);
+    }
+
+    interface NodeDiffListener {
+        NodeDiffListener NOOP = new NodeDiffListener() {
+            @Override
+            public void nodeRemoved(FileSystemNode node) {
+            }
+
+            @Override
+            public void nodeAdded(FileSystemNode node) {
+            }
+        };
+
+        void nodeRemoved(FileSystemNode node);
+        void nodeAdded(FileSystemNode node);
+    }
+
+    interface DiffCapturingUpdateFunction {
+        SnapshotHierarchy update(SnapshotHierarchy root, NodeDiffListener diffListener);
+    }
+
+    interface DiffCapturingUpdateFunctionDecorator {
+        DiffCapturingUpdateFunctionDecorator NOOP = updateFunction -> root -> updateFunction.update(root, NodeDiffListener.NOOP);
+
+        SnapshotHierarchyReference.UpdateFunction decorate(DiffCapturingUpdateFunction updateFunction);
+    }
+
+    interface SnapshotDiffListener {
+        SnapshotDiffListener NOOP = (removedNodes, addedNodes) -> {};
+
+        void changed(Collection<CompleteFileSystemLocationSnapshot> removedNodes, Collection<CompleteFileSystemLocationSnapshot> addedNodes);
```

Comment:
@lptr Do you have a better name for this `changed` method? `diff` or `snapshotDiff` would be a possibility. Or `updateFinished`. WDYF?